### DISCUSSION TITLE
Budget extensions

### DISCUSSION THREAD

**Discussion by teunbrand** [ggplot2-extenders/ggplot-extension-club#70] [Show and tell]
Have you ever come across extensions of which you think 'this could easily be implemented in vanilla ggplot2'?
I'm calling them 'budget' extensions because you're basically getting them on the cheap with ggplot2.
I've encountered tricks for [half geoms](https://github.com/teunbrand/ggplot_tricks?tab=readme-ov-file#half-geoms) before. Now I found a trick for budget `ggridges::geom_density_ridges()`:

``` r
library(ggplot2)
ggplot(ggridges::lincoln_weather, aes(x = `Mean Temperature [F]`, y = Month)) +
  geom_ribbon(
    stat = "density",
    aes(ymax = after_stat(y + scaled * 2),
        ymin = after_stat(ifelse(scaled > 0.01, y, NA)), # trims tails
        group = after_stat(-group)), # sorts correctly
    outline.type = "upper",
    colour = 'black'
  )
```

![](https://i.imgur.com/ipkhHhL.png)

<sup>Created on 2024-12-16 with [reprex v2.1.1](https://reprex.tidyverse.org)</sup>

**Comment by EvaMaeRey** [comment]
I like in-script, package free extension.  Here, I would definitely want 1) a cheap way to look it up the code to reproduce the layer (in the 'budget extensions cookbook') instead of thinking through and inspecting computed variables (no budget task for many). and 2) to wrap up the layer in-script to layer_- or geom_- density_ridges, so that it's also cheap to read back the plot.

**Reply by EvaMaeRey** [reply]
might be getting way ahead of ourselves, but probably not a bad idea to think about ways to credit what is the inspiration for the budget version.

**Comment by yjunechoe** [comment]
Unrelated, but I got an error trying out your half geom violin example - an issue of illusorily licensed scope? Seems like the data mask is set up such that `x` is somehow not discoverable from the after-scale context 🤔 

```r
library(ggplot2)
update_geom_defaults("violin", list(xmin = NULL))
p <- ggplot(mpg, aes(class, displ, colour = class)) +
  geom_violin(aes(xmin = after_scale(x)))
p
#> Error: object 'x' not found

x <- 1
p # now works


rm(list = "x")
p
#> Error: object 'x' not found
```

**Reply by yjunechoe** [reply]
But that issue aside, assuming every column in the after-scale data is up for grabs, I can extend your geom violin example to the case of **paired raincloud plots** with 100% vanilla ggplot (ex: [{ggrain}](https://github.com/njudd/ggrain))

For example, plotting the trajectory each subject undergoing pre- and post-treatment, as well as the distribution within each condition

```r
library(ggplot2)
set.seed(1)
n <- 20
df <- data.frame(
  condition = gl(2, n, labels = c("Pre", "Post")),
  subject = rep(1:n, times = 2),
  value = c(rnorm(n), rnorm(n, 3))
)
df

update_geom_defaults("violin", list(xmin = NULL, xmax = NULL))
group <- 1
xmin <- 1
xmax <- 1
x <- 1

ggplot(df, aes(condition, value, color = condition, fill = condition)) +
  geom_violin(
    aes(
      xmin = after_scale(ifelse(group == 1, xmin, x)),
      xmax = after_scale(ifelse(group == 2, xmax, x))
    )
  ) +
  geom_point(
    aes(x = stage(condition, after_scale = x + (1.5 - x)/6))
  ) +
  geom_line(
    aes(x = stage(condition, after_scale = x + (1.5 - x)/6), group = subject),
    color = "grey50"
  ) +
  theme_classic()
```

![image](https://github.com/user-attachments/assets/16e06ae7-9d31-4a27-9668-101911fff559)

**Reply by teunbrand** [reply]
I really don't feel warmly toward this type of plot, but it is cool that you found a way to reproduce it!

> an issue of illusorily licensed scope?

Yeah I'm not sure what is going on here, but you're right to point it out

**Reply by yjunechoe** [reply]
> an issue of illusorily licensed scope?

I dug into it a little further - *i think* this is specific to resolving `after_scale()` because `Layer$compute_geom_2()` (more specifically, `Geom$use_defaults()`) is called to tidyeval the aes expr against both the legend guide data and the layer data, and variables like `x` doesn't exist in the guide data:

```r
trace("compute_geom_2", tracer = quote(print(head(data))), where = ggplot2:::Layer)
#> Tracing function "compute_geom_2" in package ".GlobalEnv"
#> [1] "compute_geom_2"

p <- ggplot(diamonds, aes(clarity)) +
  geom_bar(aes(fill = stage(clarity, after_scale = fill)))
x <- ggplot_build(p)
#> Tracing compute_geom_2(..., self = self) on entry 
#>        fill .id
#> 1 #440154FF   1
#> 2 #46337EFF   2
#> 3 #365C8DFF   3
#> 4 #277F8EFF   4
#> 5 #1FA187FF   5
#> 6 #4AC16DFF   6
#> Tracing compute_geom_2(..., self = self) on entry 
#>        fill     y count prop x flipped_aes PANEL group ymin  ymax xmin xmax
#> 1 #440154FF   741   741    1 1       FALSE     1     1    0   741 0.55 1.45
#> 2 #46337EFF  9194  9194    1 2       FALSE     1     2    0  9194 1.55 2.45
#> 3 #365C8DFF 13065 13065    1 3       FALSE     1     3    0 13065 2.55 3.45
#> 4 #277F8EFF 12258 12258    1 4       FALSE     1     4    0 12258 3.55 4.45
#> 5 #1FA187FF  8171  8171    1 5       FALSE     1     5    0  8171 4.55 5.45
#> 6 #4AC16DFF  5066  5066    1 6       FALSE     1     6    0  5066 5.55 6.45

untrace("compute_geom_2", where = ggplot2:::Layer)
#> Untracing function "compute_geom_2" in package ".GlobalEnv"
```

Should I open an  issue with a reprex?

**Reply by yjunechoe** [reply]
Relatedly, the example in https://github.com/teunbrand/ggplot-extension-club/discussions/47#discussioncomment-10893280 errors once I make the layer generate a legend:

```r
library(ggplot2)
library(colorspace)

p <- ggplot(diamonds, aes(cut, group = interaction(cut, clarity))) +
  geom_bar(
    aes(fill = stage(cut,
                     after_scale = ave(fill, x, PANEL, FUN = \(.x) {
                       darken(.x, seq(0.1, 0.9, length.out = length(.x)))
                     }))),
    color = "white",
    # show.legend = FALSE
  )
p
#> Error: object 'x' not found
```

And it's not so much "illusory" licensing because the value of the global variables `x`, `PANEL`, etc. are indeed read in the case of the guides data. In other words, the `<- 1` workaround was a fluke - any constant would work to the effect of doing nothing when used to split the vector in `ave()`.

```r
x <- PANEL <- ggplot2:::NO_GROUP
p
```

![image](https://github.com/user-attachments/assets/cb1fdb54-951c-4101-a579-ded4f780557a)

Compare to the case where values of global `x` and `PANEL` would break `ave()`

```r
x <- PANEL <- NULL
p
#> Error in split.default(x, g): group length is 0 but data length > 0
```

And again, this is specific to resolving the aes expr against the guides data, so turning guides off avoids the error:

```r
x <- PANEL <- NULL
p +
  guides(fill = guide_none())
```

![image](https://github.com/user-attachments/assets/b09804de-31ae-4229-972b-e4eb05d58b24)

**Reply by teunbrand** [reply]
Good sleuthing June!

> Should I open an issue with a reprex?

I think this is a bug and probably should get fixed, so please go ahead

**Comment by EvaMaeRey** [comment]
I'm not totally up to date w/ June's observations, but for the first example, could you alternatively create a budget Stat?  

``` r
library(ggplot2)
StatBudgetRidges <- ggproto("StatBudgetRidges", 
                            StatDensity)

StatBudgetRidges$compute_panel <- function(data, scales){
  
  data <-  StatDensity$compute_panel(data, scales) 
  
  data$ymax = data$y + data$scaled * 2
  data$ymin = ifelse(data$scaled > 0.01, data$y, NA)
  data$group = -data$group
  
  data
  
}

ggplot(ggridges::lincoln_weather, 
       aes(x = `Mean Temperature [F]`, y = Month)) +
  geom_ribbon(
    stat = StatBudgetRidges,
    outline.type = "upper",
    colour = 'black'
  )
```

![](https://i.imgur.com/OiD6QVQ.png)

<sup>Created on 2025-01-02 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>

**Reply by teunbrand** [reply]
Well yes you can do that, but argueably that beats the purpose of a 'budget' extension, as you are using the fully fledged extension capabilities to make a proper extension.

**Reply by EvaMaeRey** [reply]
Ah.  I thought 'budget extension' was more about avoiding imports and low-character, in-script definitions.  I mostly just wanted to see if the alternate approach worked, borrowing your computations in the after_stat().  It's really neat. 🙌

**Reply by EvaMaeRey** [reply]
Going with the after_stat mechanism, but interested in a slightly more encapsulated layer (I like plot pipelines to be pretty readable -- if I'm honest, I don't know much about StatDensity's var 'scaled' and wouldn't have deciphered what the after_stat moves were achieving without the output plot), I think this captures a layer that can do all the anticipated things? 

``` r
library(ggplot2)

geom_budget_ridgelines <- function(mapping = list(), ...){
  
  default_aes <- aes(ymax = after_stat(y + scaled * 2),
                     ymin = after_stat(ifelse(scaled > 0.01, y, NA)), 
                     group = after_stat(-group))
  
  geom_ribbon(
    stat = "density",
    mapping = aes(!!!modifyList(default_aes, mapping)),
    ...
  )
  
}


library(ggplot2)
ggplot(data = ggridges::lincoln_weather) +
  aes(x = `Mean Temperature [F]`, y = Month) +
  geom_budget_ridgelines(outline.type = "upper", 
                         color = "white", 
                         aes(fill = Month))
```

![](https://i.imgur.com/96Y0HUn.png)

<sup>Created on 2025-01-08 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>

**Reply by teunbrand** [reply]
> I like plot pipelines to be pretty readable

Me too but it is the first thing I give up when I'm actually building a figure for a paper 😅 
I have plots where the code is just an A4 sized wall of text fussing about the smallest details.

**Reply by EvaMaeRey** [reply]
All that theme and guides work for a more readable future!  I guess at this point maybe I'd assert that *exploratory* plot pipelines can/should be readable.

Here's a budget geom_pie() w the modifyList approach - doesn't use any cool after_stat or after_scale functions, but there are just several adjustment from vanilla bar -> vanilla pie (some that need to be coordinated like x v y and which theta) that encapsulation seems worthwhile.  I'd made some previous attempts at this w/o modifyList and didn't manage, but I think this is the right track.

``` r
library(ggplot2)

geom_pie <- function(mapping = list(), coord_polar = T, ...){
  
  default_aes <- aes(y = "All")
  
  layer <- geom_bar(stat = "count", 
                    position = "fill",
                    mapping = aes(!!!modifyList(default_aes, mapping)), 
                    ...)
  
  coord <- if(coord_polar){coord_polar()}else{NULL}
  
  list(layer, coord)

}


ggplot(diamonds) + 
  geom_pie() + 
  aes(fill = cut)
```

![](https://i.imgur.com/B1azZub.png)

``` r

last_plot() + 
  facet_wrap(~color)
```

![](https://i.imgur.com/wn1ggr6.png)

<sup>Created on 2025-01-08 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>
